src/internal_rules.jl

```diff
@@ -780,54 +772,58 @@ function EnzymeRules.forward(::Const{typeof(cholesky)}, RT::Type, A; kwargs...)
     end
 end
 
-# y = inv(A) B
-# dY = inv(A) [ dB - dA y ]
-# ->
-# B(out) = inv(A) B(in)
-# dB(out) = inv(A) [ dB(in) - dA B(out) ]
-function EnzymeRules.forward(
-        func::Const{typeof(ldiv!)},
-        RT::Type,
-        fact::Annotation{<:Cholesky},
-        B;
-        kwargs...
-)
-    if isa(B, Const)
-        @assert (RT <: Const)
+function _cholesky_forward(C::Cholesky, Ȧ)
+    # Computes the cholesky forward mode update rule
+    # C.f. eq. 8 in https://arxiv.org/pdf/1602.07527.pdf
+    if C.uplo === 'U'
+        U = C.U
+        U̇ = Ȧ / U
+        ldiv!(U', U̇)
+        idx = diagind(U̇)
+        U̇[idx] ./= 2
+        triu!(U̇)
+        rmul!(U̇, U)
+        U̇ .+= UpperTriangular(Ȧ)' .- Diagonal(Ȧ) # correction for unused triangle
+        return Cholesky(U̇, 'U', C.info)
+    else
+        L = C.L
+        L̇ = L \ Ȧ
+        rdiv!(L̇, L')
+        idx = diagind(L̇)
+        L̇[idx] ./= 2
+        tril!(L̇)
+        lmul!(L, L̇)
+        L̇ .+= LowerTriangular(Ȧ)' .- Diagonal(Ȧ) # correction for unused triangle
+        return Cholesky(L̇, 'L', C.info)
+    end
+end
+
+function EnzymeRules.forward(func::Const{typeof(ldiv!)},
+                             RT::Type{<:Union{Const,Duplicated}},
+                             fact::Annotation{<:Cholesky},
+                             B::Annotation{<:AbstractVecOrMat};
+                             kwargs...)
+    if B isa Const
         return func.val(fact.val, B.val; kwargs...)
     else
         N = width(B)
-
-        @assert !isa(B, Const)
-
-        retval = if !isa(fact, Const) || (RT <: Const) || (RT <: Duplicated) || (RT <: BatchDuplicated)
-            func.val(fact.val, B.val; kwargs...)
-        else
-            nothing
-        end
+        retval = B.val
 
         dretvals = ntuple(Val(N)) do b
             Base.@_inline_meta
-
-            dB = if N == 1
-                B.dval
+            dB = N == 1 ? B.dval : B.dval[b]
+            if fact isa Const
+                ldiv!(fact.val, B.val)
+                ldiv!(fact.val, dB)
             else
-                B.dval[b]
-            end
-
-            if !isa(fact, Const)
-
-                dfact = if N == 1
-                    fact.dval
-                else
-                    fact.dval[b]
-                end
-                
-                tmp = dfact.U * retval
-                mul!(dB, dfact.L, tmp, -1, 1)
+                dfact = N == 1 ? fact.dval : fact.dval[b]
+                L = fact.val.L
+                U = fact.val.U
+                dL = dfact.L
+                dU = dfact.U
+                _ldiv_Cholesky_forward!(L, U, B.val, dL, dU, dB)
```

Comment:
Now refactor this to call `_ldiv_Cholesky_forward!(L, U, B.val, dL, dU, B.dval)` instead of executing this in the `ntuple` call (see below)

Reply:
What about batches in `dL` and `dU`?

Reply:
Done I believe.